src/fairchem/core/common/relaxation/optimizers/lbfgs_torch.py

```diff
@@ -142,102 +126,84 @@ def run(self, fmax, steps):
+    def step(self, iteration: int) -> None:
+        # cast forces and positions to float64 otherwise the algorithm is prone to overflow
+        forces = self.optimizable.get_forces(apply_constraint=True).to(
+            dtype=torch.float64
+        )
+        pos = self.optimizable.get_positions().to(dtype=torch.float64)
 
         # Update s, y, rho
         if iteration > 0:
-            s0 = r - self.r0
+            s0 = pos - self.r0
             self.s.append(s0)
 
             y0 = -(forces - self.f0)
             self.y.append(y0)
 
-            self.rho.append(1.0 / _batched_dot(y0, s0))
+            self.rho.append(1.0 / self._batched_dot(y0, s0))
 
         loopmax = min(self.memory, iteration)
-        alpha = forces.new_empty(loopmax, self.batch.natoms.shape[0])
+        alpha = forces.new_empty(loopmax, self.optimizable.batch.natoms.shape[0])
         q = -forces
 
         for i in range(loopmax - 1, -1, -1):
-            alpha[i] = self.rho[i] * _batched_dot(self.s[i], q)  # b
-            q -= alpha[i][self.batch.batch, ..., None] * self.y[i]
+            alpha[i] = self.rho[i] * self._batched_dot(self.s[i], q)  # b
+            q -= alpha[i][self.optimizable.batch_indices, ..., None] * self.y[i]
 
         z = self.H0 * q
         for i in range(loopmax):
-            beta = self.rho[i] * _batched_dot(self.y[i], z)
+            beta = self.rho[i] * self._batched_dot(self.y[i], z)
             z += self.s[i] * (
-                alpha[i][self.batch.batch, ..., None]
-                - beta[self.batch.batch, ..., None]
+                alpha[i][self.optimizable.batch_indices, ..., None]
+                - beta[self.optimizable.batch_indices, ..., None]
             )
 
         # descent direction
         p = -z
-        dr = determine_step(p)
+        dr = self.determine_step(p)
+
+        # if dr[self.optimizable.update_mask[self.optimizable.batch_indices]].isnan().any():
+        #     breakpoint()
+
         if torch.abs(dr).max() < 1e-7:
             # Same configuration again (maybe a restart):
             return
 
-        self.set_positions(dr, update_mask)
-
-        self.r0 = r
+        self.optimizable.set_positions(pos + dr)
+        self.r0 = pos
         self.f0 = forces
 
-    def write(self, energy, forces, update_mask) -> None:
-        self.batch.energy, self.batch.force = energy, forces
-        atoms_objects = batch_to_atoms(self.batch)
-        update_mask_ = torch.split(update_mask, self.batch.natoms.tolist())
-        for atm, traj, mask in zip(atoms_objects, self.trajectories, update_mask_):
-            if mask[0] or not self.save_full:
+    def write(self) -> None:
+        atoms_objects = self.optimizable.get_atoms_list()
+        for atm, traj, mask in zip(
+            atoms_objects, self.trajectories, self.optimizable.update_mask
+        ):
+            if (
+                mask or not self.save_full
+            ):  # should this be "if mask or self.save_full"?
```

Comment:
Yeah I think `save_full` here is not used at all. We should remove `save_full` from this conditional entirely. We only need mask to know if the specific atoms object hasn't been written and needs writing. `save_full` is used to trigger `write`.